### PR TITLE
[tune] Add warnings if tune event loop gets clogged

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Add some simple warnings if it looks like Tune is bottlenecked processing results.

### CODE REVIEW

File: python/ray/tune/ray_trial_executor.py [hunk 1]
```diff
@@ -275,7 +279,18 @@ def get_next_available_trial(self):
         # the first available result, and we want to guarantee that slower
         # trials (i.e. trials that run remotely) also get fairly reported.
         # See https://github.com/ray-project/ray/issues/4211 for details.
+        start = time.time()
         [result_id], _ = ray.wait(shuffled_results)
+        wait_time = time.time() - start
+        if wait_time > NONTRIVIAL_WAIT_TIME_THRESHOLD_S:
+            self._last_nontrivial_wait = time.time()
+        if time.time() - self._last_nontrivial_wait > BOTTLENECK_WARN_PERIOD_S:
```
Comment (richardliaw) [inline python/ray/tune/ray_trial_executor.py]: hm, if all `wait_time > NONTRIVIAL_WAIT_TIME_THRESHOLD_S`, then this message would never show up -- is this intended?

Comment (ericl) [reply python/ray/tune/ray_trial_executor.py]: My intuition here is that we should only warn if tune is constantly backlogged. If there was a wait that was slow, then probably it is not so.

File: python/ray/tune/ray_trial_executor.py [hunk 2]
```diff
@@ -275,7 +279,18 @@ def get_next_available_trial(self):
...
+            logger.warn(
+                "Over the last {} seconds, ".format(BOTTLENECK_WARN_PERIOD_S) +
+                "the Tune event loop has been backlogged processing new "
+                "results. Consider increasing your period of result reporting "
+                "to improve performance.")
```
Comment (hartikainen) [inline python/ray/tune/ray_trial_executor.py]:
```python
logger.warn(
    "Over the last {} seconds, the Tune event loop has been backlogged "
    "processing new results. Consider increasing your period of result "
    reporting to improve performance.".format(BOTTLENECK_WARN_PERIOD_S))
```

File: python/ray/tune/trial_runner.py [hunk 3]
```diff
@@ -229,13 +229,14 @@ def step(self):
         """
         if self.is_finished():
             raise TuneError("Called step when all trials finished?")
-        self.trial_executor.on_step_begin()
-        next_trial = self._get_next_trial()
+        with warn_if_slow("on_step_begin"):
```
Comment (hartikainen) [inline python/ray/tune/trial_runner.py]: Would I see these warnings in the application logs in between the tune results?

File: python/ray/tune/ray_trial_executor.py [hunk 4]
```diff
@@ -18,6 +18,9 @@
 
 logger = logging.getLogger(__name__)
 
+BOTTLENECK_WARN_PERIOD_S = 60
+NONTRIVIAL_WAIT_TIME_THRESHOLD_S = 0.001
```
Comment (hartikainen) [inline python/ray/tune/ray_trial_executor.py]:
```suggestion
NONTRIVIAL_WAIT_TIME_THRESHOLD_S = 1e-3
```

File: python/ray/tune/trial_runner.py (hunk 3 above)
Comment (ericl) [reply python/ray/tune/trial_runner.py]: Yeah

**Review by hartikainen — APPROVED** [review]

**Pull request merged**